src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs

```diff
@@ -3070,5 +3070,186 @@ interface IGoo<T> { public T Count { get; } }
+                {
+                    static void Main()
+                    {
+                        var x = new C<object>();
+                        _ = nameof(x[0]);
+                        _ = nameof(x[0] = default);
+                        var y = new C<int>();
+                        _ = nameof(y[0] += 1);
+                    }
+                }
+                """;
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
+                // (10,20): error CS8081: Expression does not have a name.
+                //         _ = nameof(x[0]);
+                Diagnostic(ErrorCode.ERR_ExpressionHasNoName, "x[0]").WithLocation(10, 20),
+                // (11,20): error CS8081: Expression does not have a name.
+                //         _ = nameof(x[0] = default);
+                Diagnostic(ErrorCode.ERR_ExpressionHasNoName, "x[0] = default").WithLocation(11, 20),
+                // (13,20): error CS8081: Expression does not have a name.
+                //         _ = nameof(y[0] += 1);
+                Diagnostic(ErrorCode.ERR_ExpressionHasNoName, "y[0] += 1").WithLocation(13, 20));
+        }
+
+        [Fact]
+        public void Nameof_Indexer_05()
+        {
+            string source = """
+                ref struct R<T>
+                {
+                    public T this[int i] => default;
+                }
+                class Program
+                {
+                    static void Main()
+                    {
+                        var r = new R<object>();
+                        _ = nameof(r[0]);
+                    }
+                }
+                """;
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
+                // (10,20): error CS8081: Expression does not have a name.
+                //         _ = nameof(r[0]);
+                Diagnostic(ErrorCode.ERR_ExpressionHasNoName, "r[0]").WithLocation(10, 20));
+        }
+
+        [Fact]
+        public void Nameof_Indexer_06()
+        {
+            string source = """
+                ref struct R<T>
+                {
+                    public T this[int i] { set { } }
+                }
+                class Program
+                {
+                    static void Main()
+                    {
+                        var r = new R<object>();
+                        _ = nameof(r[0]);
+                        _ = nameof(r[0] = default);
+                    }
+                }
+                """;
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
+                // (10,20): error CS8081: Expression does not have a name.
+                //         _ = nameof(r[0]);
+                Diagnostic(ErrorCode.ERR_ExpressionHasNoName, "r[0]").WithLocation(10, 20),
+                // (10,20): error CS0154: The property or indexer 'R<object>.this[int]' cannot be used in this context because it lacks the get accessor
```

Comment:
>// (10,20): error CS0154: The property or indexer 'R<object>.this[int]' cannot be used in this context because it lacks the get accessor [](http://example.com/codeflow?start=16&length=136)

This looks like a noise